decryptor/base/encoder.go

```diff
@@ -28,3 +39,193 @@ func (e *EncodingError) Is(err error) bool {
 func NewEncodingError(column string) error {
 	return &EncodingError{column}
 }
+
+// EncodingValue represents a (possibly parsed and prepared) value that is
+// ready to be encoded
+type EncodingValue interface {
+	// AsPostgresBinary returns value encoded in postgres binary format
+	AsPostgresBinary() []byte
```

Comment:
we use `base` package to store common things for all databases. what we will do after adding one more protocol? for example mysql X or mariaDB or mssql?)
okay, we want return type aware values line Int/String/Bytes which knows how to encode value according to DBs protocol. okay, there is great place for design patterns)

we can introduce fabric that can create new Int/String/Bytes values that will implement EncodingValue with methods like `EncodeToBinary`/`EncodeToInt`:
```
type TypeFactory interface {
  NewBytesValue(...) EncodingValue
  NewIntValue(...) EncodingValue
  NewStringValue(...) EncodingValue
}
```

Then update common function `EncodeDefault` and accept +1 parameter `func EncodeDefault(..., typeFactory TypeFactory)` and use it to create new Int/String values instead of calling functions `NewBytesValue` that should implement methods for mysql/postgresql/etc

After that target db package should implement own fabric that will return own implementations of `EncodedValue` and return them.

Reply:
Yeah, you are right. It was one of the possible implementations, but I thought that it has much complexity. But current implementation is hard to extend. So, good idea!